src/index.d.ts

```diff
@@ -1,966 +1,54 @@
-export as namespace Reselect;
-
-export type Selector<S, R> = (state: S) => R;
-
-export type OutputSelector<S, R, C> = Selector<S, R> & {
-  resultFunc: C;
-  recomputations: () => number;
-  resetRecomputations: () => number;
-}
+// Minimum TypeScript Version: 4.2
 
-export type ParametricSelector<S, P, R> = (state: S, props: P, ...args: any[]) => R;
+export as namespace Reselect;
 
-export type OutputParametricSelector<S, P, R, C> = ParametricSelector<S, P, R> & {
-  resultFunc: C;
+export type Selector<S = any, R = unknown, P extends never | (readonly any[]) = any[]> = [P] extends [never] ? (state: S) => R : (state: S, ...params: P) => R;
```

Comment:
My suggestion would be to keep `Selector`, `ParametricSelector` and `OutputParametricSelector` in as it they are, mark them as `@deprecated` and use a new name for this type to avoid a breaking change.

Reply:
Hmm, so I think if I address the suggestion [here](https://github.com/reduxjs/reselect/pull/480/files/66fcf9000bc173ca1f168656d2611e8448ecc188#r567287907) these _will_ be backward compatible right?

This is because the new `Selector` is just a combined version of the old `Selector` and `Parametric` Selector. The only thing that could be considered breaking, in my opinion, is the "extra args" bit of the `ParametricSelectors`. Those were always just passed through as `...any[]` where they are not fully typed. In every other instance, I think these types will behave exactly the same, no?

See this [playground link](https://www.typescriptlang.org/play?#code/PTAECpK6dqCgQXKA8gGwCagCoE8AHAUwGdQBlIgYwBcBLAewDtEw52P54iAPAhgE41QNQkTRZK6ajUEAecgBpQAJQB8oALygAFCRoBDGkQBcFAJRaNKgNxde-ISLETMqAK40CnqTPlLVZQBhDW1fWn9ldVAAMlAAb3hQUAFSd3QaADF3JiozILtk1KoGAFtvQ3pmEjMdS00NJndSgCMiAUKU0iIaFWoyiqNGJhrdesbmto74AF97PkFhUWJQAAUDAQNSnoE6KnDZAQVlVajQ3X0jUwplAgEGAlHT0AA6N42Ac1GDJjwAbQAuuNVHZuAsnMtxB4vJ51pttjRdvsiNIIkcAs8VMFznCtjs9gdImszrEEkkuiR0lkcnlQAVycUBp4htVasCmq12p1UiQen0SuVmVURmyrKAOVM7HN4JhqOgNuIAGY04WgBhYIKpK6E9FRACMUQATFEAMzKHBqHTk3mow56swYTA646qPVqRTWlF+ASGh2SL1ol0qQ3uz22wQmv1OgOHIMm0PJAUtOhMdq1Hn213KHm+1TGimR1Qm4E4D3mKPQirOgKl3QZswqA0U3PB7OkQsqYtii12WVUeWpUDK3Kq9WYTVEbUxomYputovmy1h72Z3EIpHVk76hOgG3e3Nr-HI8O64l5nd7tGFw+IgnT0+Y+MexNlZOpgTp0iZxttkgt-M8h2XYNLgZYOp4FQ3hu94us8tY6PWWbNg2AHtg2wEaD2XBIBwuEwKwyCgAAckQADuuBiGQlC0MMBF4fR4DzI4SwuNWWigD8eBROxOQANZMAwpFMCcoC8MYTCYGQqYAG7tKAAA+qiTpgzDoHgACCAibHgcicecnGAucfyrACok8OJkmgH8MntKZAD8FyVNc5DAiooC1JcxhmAEbwvAQGxbE8rl2A4izOCslY+DB5BmRZVH3pp2lRGkGQnAFpRkGJRASWQWoqUwakcb8gLBK+KbtOczoAOI9OQTmZPcpQ6iQCjukplKpWs6UkBocSJEUKXUrk+Rle+3L9IKlTDKMdRihKXIMt0vQTYMwozeykwLTMoKQhQCVaQYeDsX0Bj5WpiWHQo95qFwu01TQdVXA1ZTVpVsXZZZzopoqcmPcYGgOX94hmDZ0x3bV9WNc1CjvTle0nhdeChOSMVZXDfzfXJQTuFp2U0MovmYwIaA0AAFrZ5LJA52O40wwho5996U8koAOagZNyQz8UIwdeDMyzrOgPdQPPU1ME06kdMaIpwuQy994tez5MCDdAsC2YstPVD4s45LNCq2rINELJAjM95sOWToROgAAotIpTmICgua8YovOnbRClBoRsmztLj3WuJBu9Fb1c-D3q6b8yicco1urADawW2QTsOaD7mJz7C2hRC-s9IHwcni1ASZAwOTYNooOh+ZH3c96iPkoDSe6NbJCOwCzMOQH3UFxHPVmxQTcY0wP3ExLeME281t9Po7dq87efd9rJ5yGPUuD7Pc-zzQ+dL96LXT-joAl2XBtzxrC-wkHu+Bqv+v8+n5DkmYx8SX7Kw6tPVKI1dhdNzqiPJX0H-fa2l2LNWRskA+g9rY+UnsPOSfRSgGBTCmD4G8qYDzDjqe+DlEHIKYKg4BPNtL3wwX8PoNAcZMHwMQVqE8Xgf0Gt-ZqyUkEoKYB8NQ6DDZWQoVQmhRBWrcLMICJ+7V6bVzhjoHQvlPjfF+MCGBbdBbkCdiIgEoIZRygVEOFUwxQBUC1MYasCsiF115oAqkS5kiE2MBlERvkWG6FkQIL43l7yfwyMwhWahXKDXbuWNAEEorLycZ4w+XdL49zREXHxygZHvFcaMRhHUaDeMLr4sU4S1Cgj7AOJUejmAGKMUQExv8sEgMOpYjI1jdwKwcW8ZqAJnwGNGmmZxiS3HhzROE9Je9MkgXCWBIJMIHrRQVtUiJF9ArRMOLEjJ8SXFdJSV-XmP9+l+NSTk7CbAGK4ToigHApBhDUWFAcvZ7AuApmMAIRUBgqDiCBmSZIBgCAEGkGYfQuwOHkhaD8f5ZgWgMHVJOFgiYNj3BoCDTaptpQlBGMIS8aS3nSHYjoTARgDDeScsCDFhgXivPeUQeA8KgFIoAEL-KYAYNFeKsUUBxWKOlLw-nUupSS6oiKYxBAhQwYQ2h0WYuxVcW49xHhmHiKAFMFQzB8QEkJUAMxlDABQNS7YnzbwcIgMAXFmKXhUF5TQLgpLhB3GuXVb5Hw0XSQMOgdw1wvmoOBKSkFLx0AMA+DoG1dqiDmA5Qi0AZq6bkuBdIH41rbX2sBaG0FzrqiuvdZ6719q-UmsDd8mgREYURp9dCzkAg40jATR6r1kbfXGs5eKMiOp2KGMnMY+8VpkhIvUiiogLSKVUoMB27lhqWk6EJdIZQrL-nKANVpPlwJ+osyDQ9TVnrB3loFrOkNIKfg6BHdSv1y6M1ZvzTocdkLLDkjmH6-1QCxw1u0GOCcU4TxNrqbaVtRKe22kpWy7ty5aA8onfjckA623Dq7WOw1U7mazotaggDRLt0zozausNTAN1dtg8kWde6pgHtA6AE98Az1IBwKTOgZBiNFVEhO4m8rQBtANe4Xk4pmDiAYIqEQ5NH09EEGQe5lDbWFUMLxcQNLk1EF8h8BgoAaRlG2HTVj4h5RAP8vCNUTBQDoHKqAAAnBpoq2AyakdInQdA6BQCkUELxeAqZSI6h0PERdZgADk5M8D2aAx+swiJ7Ugd-WYE0SqEhSqYDK0A9m8DqgYPZxVyh7MtHlKTezfqkApiHNx2OwgDNGdabjGgfGXAc0DelHockDBAtksoHIamBOybVFgUAJsSD6LJkYUAAArOjwhbWkUOmQUmBgyAAANZHFQBH1jiZA8uKhTLajirjmh43CsSy9jbbNtoc05lz1Gu3uYEJ5gxhqfN+cldKzwDnQvuoi356LsX4sEQAJL2dKBJkYONxB0BY3pkjXHQAACIWhal4oQqgPWOFEC+2qYmr2pU0Hs2QVrQCaUtHcFaxrwhSO0Y+KTGgLxQA3ZY8QB4qKdF3AYNJOgmBCFIs46xpr-HxBiU2AFioJBlDvZM4Z4zsOjVICJyT2U1Geg3Pm7uAwP1RBDkELJ0ovkgA) for an example of what I'm talking about.